opalog/policy.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2016 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package opalog
+
+import "fmt"
+import "strings"
+
+// DefaultRootDocument is the default root document.
+// All package directives inside source files are implicitly
+// prefixed with the DefaultRootDocument value.
+var DefaultRootDocument = VarTerm("data")
+
+// Keywords is an array of reserved keywords in the language.
+// These are reserved names that cannot be used for variables.
+var Keywords = [...]string{
+	"package", "import", "not",
+}
+
+type (
+	// Module represents a collection of policies (defined by rules)
+	// within a namespace (defined by the package) and optional
+	// dependencies on external documents (defined by imports).
+	Module struct {
+		Package *Package
+		Imports []*Import
+		Rules   []*Rule
+	}
+
+	// Package represents the namespace of the documents produced
+	// by rules inside the module.
+	Package struct {
+		Location *Location
+		Path     Ref
+	}
+
+	// Import represents a dependency on a document outside of the policy
+	// namespace. Imports are optional.
+	Import struct {
+		Location *Location
+		Path     Value
+		Alias    Var
+	}
+
+	// Rule represents a rule as defined by Opalog. Rules define the
+	// content of documents that represent policy decisions.
+	Rule struct {
+		Location *Location
+		Name     Var
+		Key      *Term
+		Value    *Term
+		Body     Body
+	}
+
+	// Body represents one or more expressios contained inside a rule.
+	Body []*Expr
+
+	// Expr represents a single expression contained inside the body of a rule.
+	Expr struct {
+		Location *Location
+		Negated  bool
+		Builtin  bool
+		Terms    []*Term
+	}
+)
+
```

Comment:
Not sure where to put this comment...  The parser allows an Expr to be InfixExpr / PrefixExpr / Term.  InfixExpr and PrefixExpr are basically []Term.  I didn't understand why an Expr could be a term, since that would allow statements like...  p[x]=y :- q[x]=y, {1:2}, [1,2,3], r.  But it's necessary to allow an Expr to be a Term b/c of sets: p[x] :- q[x].  Is that right?   And then we use the Builtin field to cache whether the Value is a Term or []Term.  Right?